### PR TITLE
feat(connector): add support for external authentication for cybersource

### PR DESCRIPTION
**Pull request by hrithikesh026**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

 add support for external authentication for cybersource.
 After this change is merged, we'll be able to authenticate a payment through an external 3ds authentication(threedsecureio or netcetera) and authorise the same payment through cybersource.
 
Other changes:
Added `ds_trans_id` field to `authentication` table since it is required for authorization through cybersource.

### Additional Changes

- [ ] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Ran cybersource postman collection to make sure existing 3ds flow is not affected.
Screenshot 2024-05-23 at 1 00 07 PM

curls:
1. Create a cybersource connector and netcetera authentication connector.
```
curl --location 'http://localhost:8080/account/postman_merchant_GHAction_7360288d-b77e-4624-8a48-74c11c31d3a0/connectors' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '{
    "connector_type": "fiz_operations",
    "connector_name": "cybersource",
    "connector_account_details": {
        "auth_type": "SignatureKey",
        "api_key": "",
        "key1": "",
        "api_secret": ""
    },
    "test_mode": false,
    "disabled": false,
    "business_country": "US",
    "business_label": "default",
    "payment_methods_enabled": [
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "credit",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        },
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "debit",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        },
        {
            "payment_method": "wallet",
            "payment_method_types": [
                {
                    "payment_method_type": "apple_pay",
                    "payment_experience": "invoke_sdk_client",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        },
        {
            "payment_method": "wallet",
            "payment_method_types": [
                {
                    "payment_method_type": "google_pay",
                    "payment_experience": "invoke_sdk_client",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        }
    ],
    "metadata": {
        "acquirer_bin": "438309",
        "acquirer_merchant_id": "00002000000"
    }
}'
```
```
curl --location 'http://localhost:8080/account/postman_merchant_GHAction_7360288d-b77e-4624-8a48-74c11c31d3a0/connectors' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '{
    "connector_type": "authentication_processor",
    "business_country": "US",
    "business_label": "default",
    "connector_name": "netcetera",
    "connector_account_details": {
        "auth_type": "CertificateAuth",
        "certificate": "",
        "private_key": ""
    },
    "test_mode": true,
    "disabled": false,
    "metadata": {
        "mcc": "5411",
        "merchant_country_code": "840",
        "merchant_name": "Dummy Merchant",
        "endpoint_prefix": "flowbird",
        "three_ds_requestor_name": "juspay-prev",
        "three_ds_requestor_id": "juspay-prev",
        "pull_mechanism_for_external_3ds_enabled": false
    }
}'
```
2. Create a payment with "request_external_three_ds_authentication": true, and "authentication_type": "three_ds",
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_Mv6wkxquIeO6WDvY49Mpd4hullQ4CW05hBRT7tlIX4EOS5NaG5fmDzuY5NqacvXT' \
--data-raw '{
    "amount": 6540,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    "customer_id": "StripeCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "three_ds",
    "return_url": "https://duck.com",
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX"
        },
        "phone": {
            "number": "123456789",
            "country_code": "12"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX"
        },
        "phone": {
            "number": "123456789",
            "country_code": "12"
        }
    },
    "request_external_three_ds_authentication": true,
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}'
```

3. Confirm the payment
```
curl --location 'http://localhost:8080/payments/pay_XPc0gwoIZLYpMBKI8Yng/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: pk_dev_d777ba4dfb764db2a6f851870a00a7e5' \
--data '{
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "115.99.183.2"
    },
    "client_secret": "pay_XPc0gwoIZLYpMBKI8Yng_secret_vRjk6g87hvftbO9z18BO",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "card_number": "4929251897047956", 
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "card_cvc": "123"
        }
    }
}'
```

4. Authenticate the payment.
```
curl --location 'http://localhost:8080/payments/pay_XPc0gwoIZLYpMBKI8Yng/3ds/authentication' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: pk_dev_d777ba4dfb764db2a6f851870a00a7e5' \
--data '{
    "client_secret": "pay_XPc0gwoIZLYpMBKI8Yng_secret_vRjk6g87hvftbO9z18BO",
    "device_channel": "BRW",
    "threeds_method_comp_ind": "N"
}'
```

5. Authorize the payment.
```
curl --location 'http://localhost:8080/payments/pay_XPc0gwoIZLYpMBKI8Yng/postman_merchant_GHAction_7360288d-b77e-4624-8a48-74c11c31d3a0/authorize/checkout' \
--header 'Content-Type: application/json' \
--data '{
    
}'
```

6. Retrive the payment.
```
curl --location 'http://localhost:8080/payments/pay_XPc0gwoIZLYpMBKI8Yng?force_sync=true' \
--header 'Accept: application/json' \
--header 'api-key: dev_Mv6wkxquIeO6WDvY49Mpd4hullQ4CW05hBRT7tlIX4EOS5NaG5fmDzuY5NqacvXT'
```

Payment should `succeed` and below object should be returned in the payment response body.
```
"external_authentication_details": {
        "authentication_flow": "frictionless",
        "electronic_commerce_indicator": null,
        "status": "success",
        "ds_transaction_id": "3961a5b1-593f-4f94-851e-2793cd7f9e77",
        "version": "2.2.0",
        "error_code": null,
        "error_message": null
    }
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by AkshayaFoiger — DISMISSED** [review]

File: crates/router/src/connector/cybersource/transformers.rs [hunk 1]
```diff
@@ -655,6 +682,25 @@ impl
...
+        let commerce_indicator_for_external_authentication = item
+            .router_data
+            .request
+            .authentication_data
+            .as_ref()
+            .and_then(|authn_data| {
+                dbg!(&authn_data);
+                authn_data.eci.as_ref().map(|eci| {
+                    match eci.as_str() {
+                        "05" => "vbv",
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/cybersource/transformers.rs:695]: These values are specific to network of the card right?

File: crates/router/src/connector/cybersource/transformers.rs [hunk 2]
```diff
@@ -261,7 +261,34 @@ pub struct CybersourceConsumerAuthInformation {
     xid: Option<String>,
     directory_server_transaction_id: Option<Secret<String>>,
     specification_version: Option<String>,
+    /// This field specifies the 3ds version
+    pa_specification_version: Option<PaSpecificationVersion>,
+    /// Verification response enrollment status.
+    ///
+    /// This field is supported only on Asia, Middle East, and Africa Gateway.
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/cybersource/transformers.rs:268]: What about the US gateway? Should we skip sending this value or external authentication itself is not available?

File: crates/router/src/connector/cybersource/transformers.rs [hunk 3]
```diff
@@ -261,7 +261,34 @@ pub struct CybersourceConsumerAuthInformation {
     xid: Option<String>,
     directory_server_transaction_id: Option<Secret<String>>,
     specification_version: Option<String>,
+    /// This field specifies the 3ds version
+    pa_specification_version: Option<PaSpecificationVersion>,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/cybersource/transformers.rs:265]: Shouldn't we send the entire message version here?

File: crates/router/src/connector/cybersource/transformers.rs [hunk 4]
```diff
@@ -852,12 +899,42 @@ impl
...
+                        (Some(Secret::new(authn_data.cavv.clone())), None)
+                    } else {
+                        (None, Some(authn_data.cavv.clone()))
+                    };
+                CybersourceConsumerAuthInformation {
+                    ucaf_collection_indicator: None,
+                    cavv,
+                    ucaf_authentication_data,
+                    xid: Some(authn_data.threeds_server_transaction_id.clone()),
+                    // xid: None,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/cybersource/transformers.rs:919]: Can we please remove this comment?

File: migrations/2024-05-21-065403_add_ds_trans_id_to_authentication_table/down.sql [hunk 5]
```diff
@@ -0,0 +1,2 @@
+-- This file should undo anything in `up.sql`
+ALTER TABLE authentication DROP COLUMN ds_trans_id;
```
Comment (sai-harsha-vardhan) [inline migrations/2024-05-21-065403_add_ds_trans_id_to_authentication_table/down.sql:2]:
```suggestion
ALTER TABLE authentication DROP COLUMN IF EXISTS ds_trans_id;
```

File: migrations/2024-05-21-065403_add_ds_trans_id_to_authentication_table/up.sql [hunk 6]
```diff
@@ -0,0 +1,2 @@
+-- Your SQL goes here
+ALTER TABLE authentication ADD COLUMN ds_trans_id VARCHAR;
```
Comment (sai-harsha-vardhan) [inline migrations/2024-05-21-065403_add_ds_trans_id_to_authentication_table/up.sql:2]:
```suggestion
ALTER TABLE authentication ADD COLUMN IF NOT EXISTS ds_trans_id VARCHAR;
```

Comment (sai-harsha-vardhan) [inline migrations/2024-05-21-065403_add_ds_trans_id_to_authentication_table/up.sql:2]: Can we please specify some length for the VARCHAR here?

File: crates/router/src/connector/cybersource/transformers.rs [hunk 7]
```diff
@@ -655,6 +664,61 @@ impl
...
+            .as_ref()
+            .map(|card_network| card_network.to_lowercase());
+        let commerce_indicator_for_external_authentication = item
+            .router_data
+            .request
+            .authentication_data
+            .as_ref()
+            .and_then(|authn_data| {
+                authn_data.eci.as_ref().map(|eci| {
+                    match eci.as_str() {
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/cybersource/transformers.rs:678]: Can we please move this match to a fn?

**Review by sai-harsha-vardhan — DISMISSED** [review]

**Review by SamraatBansal — DISMISSED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Pull request merged**